Feature/protocol abstraction

hedefalk opened the pull request:
More typification for vscode + a first step to re-introducing old socket protocol since websockets seems problematic for some.

hedefalk committed:
wip typings

hedefalk committed:
wip ts-ify

hedefalk committed:
wip

hedefalk committed:
protocol abstraction working

hedefalk committed:
more typification

hedefalk committed:
missed chokidar saved dep + do typings install pre build

hedefalk committed:
trying fix integration test

hedefalk committed:
fixed string interpolation for ts

hedefalk committed:
fixed integration tests

hedefalk committed:
more logging

d6y commented on src/ensime-client-startup.ts:
```diff
@@ -0,0 +1,47 @@
+    if(fs.existsSync(httpPortFilePath)) {
+      // server running, no need to start
+      const httpPort = utils.removeTrailingNewline(fs.readFileSync(httpPortFilePath).toString())
+      createClient(httpPort, generalHandler).then(callback)
```
Should we be reading the `server.pid` file, an passing it to `createClient` so that we can shut down the server when we exit?  This isn't something changed in this PR (hijack!) -- I just was wondering about it is all.

hedefalk replied:
I don't think there is any server.pid file? There's a pid IFF the server was started from Atom, then we can save the handle to kill it. Or is there a server.pid file now from server? We had a discussion about this and there is a ticket on ensime-server about it but I didn't know anything had happened already…

d6y replied:
I read the comment _// server running, no need to start_ and figured there would therefore be a `servier.pid` file. Ah, so this is if it's running external to Atom, and therefore we _do not_ want to kill it. OK! Thanks,

hedefalk replied:
Well, I would actually like a way to kill it anyways - say of Atom died on you and left an orphaned Ensime server running. But It's kind of not possible without hackery right now since we don't have the pid unless we started it in this session.

Oh, we could of course write our own pid file from the client. Didn't even think about that…

rorygraves replied:
We need to bring back the -Dexplode-on-disconnect flag.  Its an ensime server flag that if the editor dies and disconnects then ensime shuts itself down.  Not sure if its currently working or if it works for websockets.

hedefalk replied:
That's a great plan! If server is started from Atom we can write a server.pid just beside server.port and use this when initializing the client so that a Atom will always have a pid if the server was started from Atom at some point in time. Then we can kill it! Yey!

hedefalk replied:
@rorygraves I'm not actually all that super-fond of that either since I want the option to have a long running Ensime server that DOESN'T die on disconnect. If it's a startup flag of course…

rorygraves replied:
Agreed,  but its a flag (which may or may not work right now)

hedefalk committed:
upgraded chokidar for later fs-events

Does this possibly fix the broken build https://drone.github.woodenstake.se/ensime/ensime-node/11 where the optional dependency on fs-events is skipped?

hedefalk committed:
error logging from coursier

hedefalk commented:
I have had failing integration tests on drone and added logging upon logging and finally it just worked again. I have no idea why, but I'm going to merge now :)
